deployment: add helm charts, configuration and openshift docs (#30)
inveniosoftware/docs-invenio-rdm: docs/deployment/index.md

# How can I deploy InvenioRDM?

You can deploy InvenioRDM in several ways. You can install it in your [local computer](../develop/index.md) to have more easily customizable environment, otherwise you can deploy a [containerized environment](../preview/index.md) that demonstrates the setup of all components runnning in docker containers. In this section it is explained how to deploy in a closer-to-production manner.

!!! warning "Do not deploy as-is in production"

>     Please note that it is mentioned as "closer-to-production", this is because even if the designed architecture can scale and withstand the load of a production service (It has been tested to stand peaks of up to 180 requests/s), the security configurations might not be enough and you should review it. In addition, it can deploy the extra services (Elasticsearch and PostgreSQL) along with them, however this ones are not configured with redundancy and persistance.

Author: I see your point. IMHO ES and DB shouldn't be deployed along with this, but for demo purposes we have to. In addition, is not about the volumes or not, is the fact that if you "helm uninstall" it all gets wiped out, so does the volume, with an external DB you assure that doesn't happen (e.g. the demo that I did on the telecon uninstall --> install and data was still there.

I think this documentation will be more clear when in the future we have the "services" properly described. There will be a more detailed needs. This is intended as a recipe, that will become able to do prod deployment, but atm we are in beta, so lets see how it evolves (along with the docs ofc). There are many corners still to be polished :)

P.S. I have made some changes and tried to adapt it better.

>     Please note that it is mentioned as "closer-to-production", this is because even if the designed architecture can scale and withstand the load of a production service (It has been tested to stand peaks of up to 180 requests/s), the security configurations might not be enough and you should review it. In addition, it can deploy the extra services (Elasticsearch and PostgreSQL) along with them, however this ones are not configured with redundancy and persistance.

Reviewer: > "helm uninstall" it all gets wiped out

Oh man, I don't want to be the person deploying things :sweat_smile: 

> more clear when in the future we have the "services" properly described. There will be a more detailed needs

Makes sense! Overall I am sure with time, the details will be added and more helpful tips will be provided.

- HAProxy as entry point. It provides load balancing and queuing of the requests.
- Nginx as reverse proxy. It serves as reverse proxy, to help HAproxy and uWSGI "talk" the same language (protocol).
- Web application nodes, running the uWSGI application.
- Redis and RabbitMQ come along in containers.
- Elasticsearch and PostgreSQL can be added to the deployment, however they are not configured in-depth and therefore not suited for more than demo purposes.

For more in-depth documentation see the [services description](services.md) and the configuration available [here](configuration.md).

## Pre-Requirements

- [Helm](https://helm.sh/docs/intro/install/) version 3.x
- Adding the [helm-invenio](https://github.com/inveniosoftware/helm-invenio) repository

``` console
$ helm repo add helm-invenio https://inveniosoftware.github.io/helm-invenio/
$ helm repo update
$ helm search invenio

NAME                   	CHART VERSION	APP VERSION	DESCRIPTION
helm-invenio/invenio	0.2.0        	1.16.0     	Open Source framework for large-scale digital repositories
helm-invenio/invenio	0.1.0        	1.16.0     	Open Source framework for large-scale digital repositories
```

> You can also install by cloning from GitHub by cloning the repository:

Reviewer: ```suggestion
You can also install by cloning the repository from GitHub:
```

```
$ git clone https://github.com/inveniosoftware/helm-invenio.git
$ cd helm-invenio/
```

> Then you will need, to reference the `./invenio` folder rather than the chart name (`helm-invenio/invenio`).

Reviewer: ```suggestion
In this case, you will need to reference the `./invenio` folder rather than the chart name (`helm-invenio/invenio`).
```

## Supported Platforms

!!! warning "Only compatible with OpenShift"

>     Pleas note that currently these Helm charts are only compatible with OpenShift.

Reviewer: ```suggestion
    Please note that currently these Helm charts are only compatible with OpenShift.
```
